### PR TITLE
storage: correct anomaly for lost delete range

### LINKED ISSUE #6240: kv: DeleteRange anomaly at SNAPSHOT
`TestTxnDBLostUpdateAnomaly_DeleteRange` currently fails with variations of

>
iso=[SNAPSHOT SNAPSHOT SNAPSHOT], pri=[3 2 1]
I1.1(A)[1] C1.1 DR3.1(A-C) R2.1(A)[1] C3.1 W2.1(B-A)[1] C2.1

and

> iso=[SNAPSHOT SNAPSHOT SNAPSHOT], pri=[2 3 1]
I1.1(A)[1] C1.1 R2.1(A)[1] DR3.1(A-C) W2.1(B-A)[1] C3.1 C2.1

In the first
* the value A=1 is written and committed
* transaction 3 deletes A-C (which means leaving an intent on A and
  populating the write timestamp cache interval A-C.
* transaction 2 pushes the intent on A and reads A=1.
  The write timestamp cache entry on A isn't updated properly in the process
  (which is an unrelated issue that likely incurs anomalies).
* transaction 2 writes B=1, commits.
  This is the anomaly: logically, there is DeleteRange intent sitting there
  with a higher timestamp, so this should be a write-write conflict. But there
  isn't an actual intent there (since the key was previously vacant, just like
  the countably infinite rest of vacant keys around it).
* transaction 3 commits (at its higher timestamp), so now A=<nil>,
  but note that we still have B=1 (written by transaction 2).

Note that this example should fail in the same way with only transaction 3 at
SNAPSHOT isolation.

The one obvious fix that comes to mind is to "somehow" push the whole range
entry in the timestamp cache forward along with the push of one of its intents.
However, that seems brittle and may not reliably be possible since the
timestamp cache's current implementation freely splits up its intervals as new
entries are added.

In the second one, the read of A and the range deletion are interchanged:

```
iso=[SNAPSHOT SNAPSHOT SNAPSHOT], pri=[2 3 1]
I1.1(A)[1] C1.1 R2.1(A)[1] DR3.1(A-C) W2.1(B-A)[1] C3.1 C2.1
```

In other words:

* write and commit A=1 as before
* txn2 reads A=1
* txn3 deletes A-C (so `ts(txn3) > ts(txn2)`)
* txn2 writes B=1 (pushing its timestamp due to tsCache entry on A-C); now
  again `ts(txn3) < ts(txn2)`
* both commit and the visible values are A=<nil>, B=1; we've again lost txn3's
  deletion of B.

This seems a little less anomalous than the above since essentially what
happens is that txn2 has write skew: It reads a stale value of A and writes
it somewhere, except that this "somewhere" would be a write-write conflict in
most monolithic databases.

See https://github.com/cockroachdb/cockroach/pull/6227 for the PR that introduced these tests.

### PR DESCRIPTION
**Pull request by spencerkimball**
Take note of DeleteRange and set a transaction flag which requires the transaction is retried in the event it's commit timestamp is pushed forward from the original timestamp.

Fix correctness tests for LostDelete & LostDeleteRange.

Fixes #6240

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/6548)

### CODE REVIEW

**Comment by bdarnell** [comment]
Commit messages should generally start with the package name. The commit message should place more emphasis on the change to production code, not the tests.

---

Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[roachpb/data.proto, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-KH29i61Sx7Qh6aIIsk6:-KH29i628JnZEWw8GrnR:1209836588) ([raw file](https://github.com/cockroachdb/cockroach/blob/978235e8f8be9916451ad8d5093fcc21a81bf539/roachpb/data.proto#L298)):*
We can't make any changes to this proto until the new migration system is in place; that's what the below_raft_protos_test is for.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-:-KH29MKVuF9F78IBx2-V:1026981682)*

File: kv/txn_correctness_test.go [hunk 1]
```diff
@@ -765,60 +772,68 @@ func TestTxnDBLostUpdateAnomaly(t *testing.T) {
...
+// TestTxnDBLostDeleteRangeAnomaly verifies that neither SI nor SSI
+// isolation are subject to the lost delete range anomaly. See #6240.
+//
+// With lost delete range, the delete range for keys B-C leave no
+// deletion tombstones (as there are an infinite number of keys in the
+// range [B,C)). Without deletion tombstones, the anomaly manifests in
+// snapshot mode when txn1 pushes txn2 to commit at a higher timestamp
+// and then txn1 writes B and commits an an earlier timestamp. The
+// delete range request therefore committed but failed to delete the
+// value written to key B.
```
Comment (tschottdorf) [inline kv/txn_correctness_test.go]: > // This anomaly is prevented by making snapshot transactions which involve a range deletion restart when they are pushed.

File: storage/replica.go [hunk 2]
```diff
@@ -1028,6 +1028,11 @@ func (r *Replica) endCmds(cmd *cmd, ba *roachpb.BatchRequest, br *roachpb.BatchR
 					// DeleteRange adds to the write timestamp cache to prevent
 					// subsequent writes from rewriting history.
 					readTSCache = false
+					// DeleteRange requires that we retry on push to avoid the
+					// lost delete range anomaly.
+					if br.Txn != nil {
+						br.Txn.RetryOnPush = true
```
Comment (tschottdorf) [inline storage/replica.go]: Can you move this into a txn returned by `DeleteRange` itself? This is an odd place.

**Comment by tschottdorf** [comment]
I think this is a reasonable fix (especially since we don't want to get sidetracked on this for now).
@bdarnell, is this really blocked by the migration story? Do we support downgrades at the moment? Commands proposed with the old version of Cockroach will never have the flag set, so its application will proceed the same with the old or new.

File: kv/txn_correctness_test.go (hunk 1 above)
Comment (spencerkimball) [inline kv/txn_correctness_test.go]: Done

**Comment by spencerkimball** [comment]
Fixed the commit title.

So should I just wait on this?

@tschottdorf, yes the tests failed in exactly the same way after the refactoring. The refactoring I did was to eliminate unnecessary permutations caused by having three txns.

---

Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[roachpb/data.proto, line 298 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-KH29i61Sx7Qh6aIIsk6:-KH5r1Sr2PCU1LAQi50g:1909930815) ([raw file](https://github.com/cockroachdb/cockroach/blob/978235e8f8be9916451ad8d5093fcc21a81bf539/roachpb/data.proto#L298)):*
I figured there was something I wasn't getting about that test. How do we proceed? What's ETA for new migration system?

---

*[storage/replica.go, line 1034 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-KH1yb4P9D0EFtLO2Z-i-r1-1034:-KH5rGpumVUX9VLHTl4U:-683378650) ([raw file](https://github.com/cockroachdb/cockroach/blob/978235e8f8be9916451ad8d5093fcc21a81bf539/storage/replica.go#L1034)):*
It's a little odd here, but then again, it's somewhat symmetric with setting WriteTooOld when consulting the timestamp cache. The reason I don't do it in the `DeleteRange` itself is that command doesn't return a transaction with the reply. I suppose I could add it there, but not sure I love that idea.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-:-KH5qoX1j7vxQUY1U1r1:-1856029574)*

**Comment by tschottdorf** [comment]
Let's see what @bdarnell says - I hope this is a case that doesn't matter unless we're not supporting downgrades (even if we did, the freeze/unfreeze stuff should be good to go soon, hoping next week).

You should also tag this reference for Jesse because it's a change in semantics - snapshot transactions which involve a DeleteRange under the hood will behave more like serializable transactions. I suppose the best way to put that in  docs is that we don't guarantee that snapshot transactions really run as snapshot (This is similar to other isolation levels which we support only by upgrading to the next allowed level, which in itself I'm not sure is properly documented). And that leads to another TODO (which should be filed) that we should find a way of exposing the "actual" mode in one of the `EXPLAIN (XXX)` queries (for users who want to make sure that the specific query they're running does benefit from SNAPSHOT).

---

Reviewed 10 of 10 files at r1.
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1034 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-KH1yb4P9D0EFtLO2Z-i-r1-1034:-KH69Aqwx_3FxCyeGJJD:-1821823008) ([raw file](https://github.com/cockroachdb/cockroach/blob/978235e8f8be9916451ad8d5093fcc21a81bf539/storage/replica.go#L1034)):*
I have sympathy for not wanting to clone a txn for nothing, but please move it there anyway. With leader-proposed Raft it will have to go there (or anywhere that's closer to there than here) anyway.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-:-KH68LSB-Zu9pO3xKLjX:171876673)*

**Comment by bdarnell** [comment]
True, a stop-the-world upgrade for this is sufficient, so it doesn't need to wait for migrations. 

I don't think we need to document this. The semantics of isolation levels are the anomalies that are guaranteed not to occur; higher isolations levels always satisfy the semantics of lower ones. The database is allowed to optimize whatever it can while still preventing those anomalies, which often results in lower isolation levels performing better, but there are no guarantees of this. Besides, if we document this interaction of snapshot isolation and DeleteRange, we'll also have to document exactly which operations result in DeleteRange, etc.

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-:-KH6LnjTKKTgxRnVA5H_:1981224111)*

**Comment by spencerkimball** [comment]
@tschottdorf I don't think this needs to be documented. We already will restart snapshot transactions on write-too-old errors as well. I really can't fathom an addendum to the documentation which will be helpful at this point as `DeleteRange` is not even exposed in any useful way to end users.

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1034 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-KH1yb4P9D0EFtLO2Z-i-r1-1034:-KH6XBw98Jv6KbiHpPzk:1098408399) ([raw file](https://github.com/cockroachdb/cockroach/blob/978235e8f8be9916451ad8d5093fcc21a81bf539/storage/replica.go#L1034)):*
OK, changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-:-KH6PftrwbVRUH6SIlJz:1551318726)*

**Comment by tschottdorf** [comment]
Just for the record, I wasn't asking for these details to be included. But
I'm ok with punting, @bdarnell's explanation makes sense (though eventually
I think it still does make sense that specifying general statements like
that READ UNCOMMITTED in Cockroach really doesn't make anything faster).

> --

-- Tobias

**Comment by tschottdorf** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6548#-:-KH7UO4C5r-LbuVpXuL1:1417088721)*

**Pull request merged**